_posts/2022-06-18-zio-streams.md

```diff
@@ -0,0 +1,537 @@
+number of addition ones that are available directly on `ZStream` and can be
+referenced in the
+[official docs](https://zio.dev/next/datatypes/stream/zstream/#operations) .
+
+## Handling Failures
+
+If we think of some `val s1: ZStream[R, E, A]`, we can use the `orElse` methods
+to directly provide a recovery stream. The interesting thing of `orElseEither`
+is that we can distinguish the transition when `s1` failed, and `s2` was used,
+as `s1` values would be a `Left` and values from `s2` would be `Right`.
+
+```scala
+s1.orElse(s2: => ZStream[R1, E1, A1]): ZStream[R1, E1, A1]
+s1.orElseEither(s2: => ZStream[R1, E1, A1]): ZStream[R1, E1, Either[A, A1]]
+```
+
+If we want to recover from specific failures via
+`PartialFunction[E, ZStream[R1, E1, A1]]`s we can use `catchSome` ( or
+`catchSomeCause` for causes - a.k.a _defects_).
+
+```scala
+s1.catchSome {
+  case e1: Error1 => s2
+}
+```
+
+If we want to exhaustively recover from errors via `E => ZStream[R1, E2, A1]`)
+(and causes), we can use `catchAll` (or `catchAllCause`).
+
+```scala
+s1.catchAll {
+  case e1: Error1 => s2
+  case e2: Error2 => s3
+  case _          => s4
+}
+```
+
+If we want to push our error handling downwards, we can also transform a
+`ZStream[R, E, A]` to `ZStream[R, Nothing, Either[E, A]]` via `either` (e.g.
+`s1.either`).
+
+Recovering form a failure in a `ZStream` generally revolves around providing a
+secondary stream to recover with. If you are looking to recover "in-place", by
+effectively just dropping that element which caused trouble while processing the
+data, you could do something heavy handed like `s1.either.collectRight`. For
+more robust solutions, we would likely need to encode that logic elsewhere.
+
+## Async ZStreams
+
+So far, we've discussed fairly bounded streams, from defined/known collections
+of values. It won't be long before we want to "feed" data into a stream we've
+previously created. Luckily, we can create a `ZStream` from a `Queue`/`Hub`
+directly, which are asynchronous data structures in ZIO. In our example, we'll
+use a `Queue`, and with a reference to that `queue` we can offer values to it
+elsewhere. Additional, we can also create a `ZSink` from a `Queue`/`Hub`!
+
+If we can set both the _endpoint_ and the _source_ as a queue, what if we joined
+two `ZStream`s together with the _same_ queue? Let's look at a toy example,
+where we "process" one stream to sanitize values, that we then feed that into
+another stream that now doesn't have to address that concern.
+
+After our `ZStream` practice above, we might first write something like this:
+
+```scala
+  // Bad code!
+  val program: ZIO[Any, Throwable, ExitCode] = for {
+    queue <- Queue.unbounded[Int]
+    producer <- dirtyStream.via(parsePipeline)
+      .run(ZSink.fromQueue(queue))
+    result <- ZStream.fromQueue(queue)
+      .run(ZSink.sum[Int]).debug("sum")
+  } yield ExitCode.success
+```
+
+and it will _almost_ do what we think! It will process the values, and feed them
```

Comment:
What do we think?
Reading this, I'm not sure what is supposed to happen - we can explain this to the reader. This may feel dumbing down because as the writer, we have it crystal clear, but I've never been told I explained too much, only too little.

Reply:
I tried to fill this out a bit more, to highlight the difference of what actually happened as describe by the follow up text.